### PR TITLE
Fix some failing AD tests.

### PR DESCRIPTION
**Pull request by jppelteret**
This targets issues picked up by compiling with -Werror.

### CODE REVIEW

**Comment by jppelteret** [comment]
Most of the tests now pass locally with `-Werror`:
```
$ ctest -R "adolc|sacado"
...
99% tests passed, 1 tests failed out of 300

Total Test time (real) = 287.36 sec

The following tests FAILED:
	1409 - sacado/helper_scalar_coupled_4_components_01_4.debug (Failed)
```

This last one is another diabolical bug, which is best left for another day:
```
$ ctest -R sacado/helper_scalar_coupled_4_components_01_4.debug  -V
...
1409: Test sacado/helper_scalar_coupled_4_components_01_4.debug: RUN
1409: ===============================   OUTPUT BEGIN  ===============================
1409: Built target helper_scalar_coupled_4_components_01_4.debug
1409: Generating helper_scalar_coupled_4_components_01_4.debug/output
1409: sacado/helper_scalar_coupled_4_components_01_4.debug: BUILD successful.
1409: sacado/helper_scalar_coupled_4_components_01_4.debug: RUN failed. ------ Return code 134
1409: sacado/helper_scalar_coupled_4_components_01_4.debug: RUN failed. ------ Result: /Users/jp/Programs/GitHub/deal.II/dealii/build/tests/sacado/helper_scalar_coupled_4_components_01_4.debug/failing_output
1409: sacado/helper_scalar_coupled_4_components_01_4.debug: RUN failed. ------ Partial output:
1409: J
1409: sacado/helper_scalar_coupled_4_components_01_4.debug: RUN failed. ------ Additional output on stdout/stderr:
1409:
1409: *** Test variables: Tensor + Vector + Scalar (coupled), dim = 2, Type code: 5
1409: Recorded data...
1409: independent variable values: 1.0864,2.1664,2.2464,1.01375,0.12375,0.23375,1.34375,0.275,0.55,0.57
1409: st_ad: 0x7f999201a420 0x7f999201a380 0x7f999201a2e0 0x7f999201a240
1409: t_ad: 0x7f999201b608 0x7f999201b5b8 0x7f999201b568 0x7f999201b518
1409: v_ad: 0x7f999201b468 0x7f999201b418
1409: s_ad: 0x7f999201b348
1409: psi: 0x7f999201a5d0
1409:
1409: independent variable values: 1.0864,2.1664,2.2464,1.01375,0.12375,0.23375,1.34375,0.275,0.55,0.57
1409: psi: 0.202336
1409: Dpsi:
1409: -3.256e-01 -1.633e-01 3.376e-01 4.078e-01 -7.095e-02 -3.756e-02 3.077e-01 8.829e-01 1.766e+00 7.809e-01
1409: D2psi:
1409:        nan        nan        nan        nan        nan        nan        nan        nan        nan        nan
1409:        nan        nan        nan        nan        nan        nan        nan        nan        nan        nan
1409:        nan        nan        nan        nan        nan        nan        nan        nan        nan        nan
1409:        nan        nan        nan        nan        nan        nan        nan        nan        nan        nan
1409:        nan        nan        nan        nan        nan        nan        nan        nan        nan        nan
1409:        nan        nan        nan        nan        nan        nan        nan        nan        nan        nan
1409:        nan        nan        nan        nan        nan        nan        nan        nan        nan        nan
1409:        nan        nan        nan        nan        nan        nan        nan        nan        nan        nan
1409:        nan        nan        nan        nan        nan        nan        nan        nan        nan        nan
1409:        nan        nan        nan        nan        nan        nan        nan        nan        nan        nan
1409: extracted Dpsi (st): -0.325573 0.337596 0.337596 -0.163268
1409: extracted Dpsi (t): 0.407845 -0.070946 -0.0375596 0.307686
1409: extracted Dpsi (v): 0.882923 1.76585
1409: extracted Dpsi (s): 0.780948
1409: extracted Dpsi (t): 0.407845 -0.070946 -0.0375596 0.307686
1409: extracted Dpsi (v): 0.882923 1.76585
1409: extracted Dpsi (s): 0.780948
1409: extracted D2psi (st,st): nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
1409: extracted D2psi (st,t): nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
1409: extracted D2psi (st,v): nan nan nan nan nan nan nan nan
1409: extracted D2psi (st,s): nan nan nan nan
1409: extracted D2psi (t,st): nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
1409: extracted D2psi (t,t): nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
1409: extracted D2psi (t,v): nan nan nan nan nan nan nan nan
1409: extracted D2psi (t,s): nan nan nan nan
1409: extracted D2psi (v,st): nan nan nan nan nan nan nan nan
1409: extracted D2psi (v,t): nan nan nan nan nan nan nan nan
1409: extracted D2psi (v,v): nan nan nan nan
1409: extracted D2psi (v,s): nan nan
1409: extracted D2psi (s,st): nan nan nan nan
1409: extracted D2psi (s,t): nan nan nan nan
1409: extracted D2psi (s,v): nan nan
1409: extracted D2psi (s,s): nan
1409:
1409:
1409: --------------------------------------------------------
1409: An error occurred in line <606> of file </Users/jp/Programs/GitHub/deal.II/dealii/tests/sacado/../ad_common_tests/helper_scalar_coupled_4_components_01.h> in function
1409:     void test_symmetric_tensor_tensor_vector_scalar_coupled() [dim = 2, number_t = double, ad_type_code = dealii::Differentiation::AD::NumberTypes::sacado_rad_dfad]
1409: The violated condition was:
1409:     std::abs( (d2psi_dst_x_dst - func::d2psi_dst_x_dst(st, t, v, s)).norm()) < tol
1409: Additional information:
1409:     No match for second derivative.
1409: --------------------------------------------------------
```
This last one segfaults on GCC, but clang shows that some of the results of differential computations are utter garbage. These are tough to figure out. There's likely an illegal operation being called for `Rad<Fad>` that is perfectly legitimate for `Fad<Fad>`, but I've not yet thought about what exactly it could be.

File: tests/ad_common_tests/step-44-helper_scalar_01.h [hunk 1]
```diff
@@ -2107,8 +2107,9 @@ namespace Step44
           std::cout << " SLV " << std::flush;
           if (parameters.type_lin == "CG")
             {
-              const int solver_its = tangent_matrix.block(u_dof, u_dof).m() *
-                                     parameters.max_iterations_lin;
+              const int solver_its = static_cast<int>(
+                static_cast<double>(tangent_matrix.block(u_dof, u_dof).m()) *
+                parameters.max_iterations_lin);
```
Comment (bangerth) [inline tests/ad_common_tests/step-44-helper_scalar_01.h]: Why does this make sense? Isn't the original product between two integers? Or do you get overflow?

File: tests/ad_common_tests/step-44-helper_scalar_01.h [hunk 2]
```diff
@@ -2189,8 +2190,9 @@ namespace Step44
             preconditioner_K_Jp_inv.use_matrix(
               tangent_matrix.block(J_dof, p_dof));
             ReductionControl solver_control_K_Jp_inv(
-              tangent_matrix.block(J_dof, p_dof).m() *
-                parameters.max_iterations_lin,
+              static_cast<int>(
+                static_cast<double>(tangent_matrix.block(J_dof, p_dof).m()) *
+                parameters.max_iterations_lin),
```
Comment (bangerth) [inline tests/ad_common_tests/step-44-helper_scalar_01.h]: same here and below

File: tests/ad_common_tests/step-44-helper_scalar_01.h (hunk 1 above)
Comment (jppelteret) [reply tests/ad_common_tests/step-44-helper_scalar_01.h]: `parameters.max_iterations_lin` is a double

To be honest, I should have simplified this entire test before writing it, but I was super lazy.

**Comment by jppelteret** [comment]
> I'm also unsure why the various tensors have to be explicitly initialized -- shouldn't the default constructor do this?

I also don't understand this, but here's an example of the error I get without doing so:
```
sacado/../ad_common_tests/physics_functions_03.h:51:47: error: default initialization of an object of const type 'const Tensor<1, 2, ADNumberType>' (aka 'const Tensor<1, 2, Sacado::Fad::DFad<double> >') without a user-provided default constructor
1491:   const Tensor<1, dim, ADNumberType>          N;
1491:                                               ^
1491:                                                {}
``

File: tests/ad_common_tests/step-44-helper_scalar_01.h (hunk 1 above)
Comment (bangerth) [reply tests/ad_common_tests/step-44-helper_scalar_01.h]: I tend to think that that's the bug: `max_iterations_lin` ought to be an integer of some kind.

**Comment by bangerth** [comment]
Hm, the error makes no sense: The class does have a default constructor:
```
  /**
   * Constructor. Initialize all entries to zero.
   *
   * @note This function can also be used in CUDA device code.
   */
  DEAL_II_CUDA_HOST_DEV
  Tensor() = default;
```
But who cares, it's in the test suite. I'm just wondering whether we're missing something that also needs to happen in other contexts...

File: tests/ad_common_tests/step-44-helper_scalar_01.h (hunk 1 above)
Comment (jppelteret) [reply tests/ad_common_tests/step-44-helper_scalar_01.h]: I think that `max_iterations_lin` was a mistake in the first place: I believe that systems preconditioned using Gauss-Seidel methods should converge in `n_rows` iterations or less. I don't know  why I added this parameter. I'm going to remove it now (in the tests, not step-44 itself).

**Comment by jppelteret** [comment]
I've now removed the `max_iterations_lin` parameter.

**Comment by jppelteret** [comment]
> Hm, the error makes no sense: The class does have a default constructor

Maybe its got to do with the `ADNumberType`. I'm not interested in looking into this further though. Gotta wrap this and other things up before the release...

**Pull request merged**